### ISSUE TITLE
Custom message when running post-processing script

### ISSUE THREAD

**Issue by Sonic3R** [sabnzbd/sabnzbd#2567]
Hi,

I have python script that creates a bdinfo for bluray disc and upload to google drive.
For now, the sabnzb web interface, shows status like _Running script: myscript.py_

It is possible to display instead another status like: _Creating bdinfo_ or _Uploading to GDrive_ instead of standard message ?

Would be nice to have such feature.

**Comment by sanderjo** [comment]
https://forums.sabnzbd.org/viewtopic.php?p=129540#p129540

**Comment by Sonic3R** [comment]
So is not possible or implementable to display status progress of the script ?

**Comment by sanderjo** [comment]
> So is not possible or implementable to display status progress of the script ?

I guess it is, but that is up to you! Hint: you should start working at these locations:
https://github.com/sabnzbd/sabnzbd/blob/c3c47507e72b28071da8f86e177f52aa73aa290a/sabnzbd/postproc.py#L543
https://github.com/sabnzbd/sabnzbd/blob/c3c47507e72b28071da8f86e177f52aa73aa290a/sabnzbd/newsunpack.py#L192

**Comment by Sonic3R** [comment]
I'd like an help since I'm not familiar with python

**Comment by Safihre** [comment]
It's not updated live, only after the script has fully finsihed.
We could do that by reading stdout continously, like we do for unrar/par2, just more work.

**Comment by Safihre** [comment]
Oh wait, you are right @mnightingale, I already implemented this 🙃 
So what do you mean @Sonic3R ? As long as your script prints something while it's running, we will update the status.
For example:

![image](https://github.com/sabnzbd/sabnzbd/assets/5703454/fdda3ec1-263e-4041-9d6f-5fdac493c58c)

**Comment by Sonic3R** [comment]
Can give an example of script with printing my ? I tried using print for every step but still displays default message until is finished and there will print last line

**Comment by Safihre** [comment]
Simply took the example script from our wiki, and added a loop that prints every second:
https://sabnzbd.org/wiki/scripts/post-processing-scripts
```python
import sys
import time
try:
    (scriptname, directory, orgnzbname, jobname, reportnumber, category, group, postprocstatus, url) = sys.argv
except:
    print("No commandline parameters found")
    sys.exit(1)

# continue script
for x in range(100):
    print(x)
    time.sleep(1)

# Your code goes here

# Success code
sys.exit(0)
```

**Comment by Sonic3R** [comment]
To be clear, I have the following test:

```
import sys
import time

try:
    (scriptname, directory, orgnzbname, jobname, reportnumber, category, group, postprocstatus, url) = sys.argv
except:
    print("No commandline parameters found")
    sys.exit(1)

print("step 1\n")
time.sleep(10)
print("step2\n")
time.sleep(20)

print("done")
sys.exit(0)

```

The problem that on sabnzb, does not print `step 1`, then `step 2`, but `Running script: test.py` instead. And at end of script, is displayed `done`.

**Comment by Safihre** [comment]
What version of SABnzbd are you using?

**Comment by Sonic3R** [comment]
I'm using: `4.0.1 [[1263068](https://github.com/sabnzbd/sabnzbd/commit/1263068)]`

**Comment by Safihre** [comment]
I'm not sure what is wrong on your system/script.. 
I cannot reproduce it, and the code to show every output has been inside SABnzbd for 4 years already. So nothing more I can add.

**Comment by Sonic3R** [comment]
I'm using dockered version of sabnzb on linux, if it matters.

See the screen recording, maybe will understand better:
https://github.com/sabnzbd/sabnzbd/assets/6311947/74661cd5-ac16-4d27-83da-97f238c72c24

**Comment by Safihre** [comment]
Strange. Maybe it's a buffering issue where Python behaves differently on Linux vs Windows.
Are you able to modify the source files? Probably not since you're on Docker.

**Comment by Sonic3R** [comment]
Yes, I tested the py file in cmd line. in your screen I see no live update of status ... same issue as me, only final print is displayed on sab web interface

**Comment by Safihre** [comment]
I did not create a video, just a screenshot of an intermediate value (it counts from 0 to 100, see script above), but it is updated every second.

**Comment by Sonic3R** [comment]
Even with your script is not working...Displays only last message (9):

https://github.com/sabnzbd/sabnzbd/assets/6311947/a021859b-50df-4eea-9f51-588f1fb55d68

Are you use custom build or same version as mine ?

**Comment by Safihre** [comment]
What Docker image do you use? 
If you use linuxserver you could switch to the nightly tag to checkout these changes.

**Comment by Sonic3R** [comment]
1. Yes I use that
2. I installed separately in other machine with no docker version of sabnzb
How to proceed ? change myself `misc.py` file (from `/usr/share/sabnzbdplus/`) and then restart the service ?

**Comment by Sonic3R** [comment]
I changed misc.py and same issue, no live update

**Comment by thezoggy** [comment]
> 

did you restart sab after making changes to your script / code ?

**Comment by Sonic3R** [comment]
I downloaded latest development git branch and I ran: `py SABnzbd.py -f sabnzb.ini -s 127.0.0.1:12345`, same thing

**Comment by sanderjo** [comment]
I tried with the above script, with github version of sabnzbd, and it doesn't work for me either:

From the CLI, the script works:
```
sander@zwart2204:~/.sabnzbd/sab_scripts$ ./script_with_intermediate_output.py directory orgnzbname jobname reportnumber category group  postprocstatus url
./script_with_intermediate_output.py directory orgnzbname jobname reportnumber category group postprocstatus url
0
1
2
3
4
```
... so let's use it as pp-script:


![image](https://github.com/sabnzbd/sabnzbd/assets/1273502/d1bf38fe-9f4e-48c1-b28c-d1e3745e2499)

but during running that script, nothing appears:

![image](https://github.com/sabnzbd/sabnzbd/assets/1273502/a9d0c3d4-38f6-4fea-b661-1b92ea6d9117)

... only after finishing the script, the "99" appears:

![image](https://github.com/sabnzbd/sabnzbd/assets/1273502/9bfff4b2-a76a-4efe-8591-a8a946a64a54)

**Comment by Safihre** [comment]
@sanderjo can you run SABnzbd like this and try again?
PYTHONUNBUFFERED=1 python3 SABnzbd.py

**Comment by sanderjo** [comment]
> @sanderjo can you run SABnzbd like this and try again?
> PYTHONUNBUFFERED=1 python3 SABnzbd.py

Yes, that works!

![image](https://github.com/sabnzbd/sabnzbd/assets/1273502/39803297-17bd-49de-9a06-30395281dc63)

**Comment by Sonic3R** [comment]
> I've tested @Safihre script from [#2567 (comment)](https://github.com/sabnzbd/sabnzbd/issues/2567#issuecomment-1556604463)
> 
> I worked on Windows and Mac but **not** Linux.
> 
> A quick Google revealed others with the problem but I couldn't find a solution.

It didn’t work for me on both windows and Linux 


I have python 3 installed btw , on both system.

**Comment by Safihre** [comment]
So, the problem here is buffering by the process that we call, in this case Python.
It seems it doesn't send out the output right away but buffers is.
We tried the suggestion that is documented by Python which works for @sanderjo but not for you.
So what more can we do? I am giving up.

**Comment by mnightingale** [comment]
I can’t test right now, but someone could also try `print(x, flush=True)` instead of the PYTHONUNBUFFERED env, also maybe the script could be called with `python -u script.py`

**Comment by sanderjo** [comment]
> someone could also try `print(x, flush=True)` instead of the PYTHONUNBUFFERED env

Wow, that works when I put it into the pp-script

```
for x in range(100):
    print(x, flush=True)
    time.sleep(1)
```

![image](https://github.com/sabnzbd/sabnzbd/assets/1273502/7dfc570f-ff57-49be-bdc8-d7394acc07e2)

**Comment by thezoggy** [comment]
had to look this up,
https://stackoverflow.com/questions/59812009/what-is-the-use-of-pythonunbuffered-in-docker-file

**Comment by Sonic3R** [comment]
> > someone could also try `print(x, flush=True)` instead of the PYTHONUNBUFFERED env
> 
> Wow, that works when I put it into the pp-script
> 
> ```
> for x in range(100):
>     print(x, flush=True)
>     time.sleep(1)
> ```
> 
> ![image](https://user-images.githubusercontent.com/1273502/240029956-7dfc570f-ff57-49be-bdc8-d7394acc07e2.png)
> 
> So ... no change needed in/with SAB, just in the pp-script. Nice.

Finally, it works !!!
Many thanks guys for involving in discussion and debates.

**Comment by Safihre** [comment]
Great, I was about to give up 😂
Maybe I'll still add the environment variable by default since this can still be difficult for users if they are not the authors of the script.

**Comment by Sonic3R** [comment]
Maybe can point in documentation about that flag in print. better than changing code that may lead to unexpected behaviour.

Good luck
